src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs

```diff
@@ -1085,291 +1047,200 @@ internal static string GetExpandedEnvironmentVariable(string name, EnvironmentVa
+                }
+                else
+                {
+                    s_modulePath.Add(sharedModulePath);
+                }
+            }
         }
 
         /// <summary>
-        /// Checks if $env:PSModulePath is not set and sets it as appropriate. Note - because these
-        /// strings go through the provider, we need to escape any wildcards before passing them
-        /// along.
+        /// Get all standard module paths.
+        /// Note - because these strings go through the provider, we need to escape any wildcards before passing them along.
         /// </summary>
-        private static string SetModulePath()
+        internal static IEnumerable<string> GetModulePath()
         {
+            // From %PSModulePath% environment variable
             string currentModulePath = GetExpandedEnvironmentVariable(Constants.PSModulePathEnvVar, EnvironmentVariableTarget.Process);
-            string allUsersModulePath = PowerShellConfig.Instance.GetModulePath(ConfigScope.AllUsers);
-            string personalModulePath = PowerShellConfig.Instance.GetModulePath(ConfigScope.CurrentUser);
 
-            string newModulePathString = GetModulePath(currentModulePath, allUsersModulePath, personalModulePath);
+            // Although Windows is case-insensetive we can be case-sensetive in the case.
+            if (string.Equals(s_psModulePathEnvVar, currentModulePath, StringComparison.Ordinal) &&
+                string.IsNullOrEmpty(InternalTestHooks.TestWindowsPowerShellPSHomeLocation) &&
+                InternalTestHooks.TestPowerShellPSModulePaths?.Length == 0)
+            {
+                // %PSModulePath% is the same so return the cache.
+                return s_modulePathCache;
+            }
 
-            if (!string.IsNullOrEmpty(newModulePathString))
+            lock (s_syncPSModulePath)
             {
-#if !UNIX
-                // If on Windows, we want to add the System32 Windows PowerShell module directory
-                // so that Windows modules are discoverable
-                string windowsPowerShellModulePath = GetWindowsPowerShellPSHomeModulePath();
-                if (!newModulePathString.Contains(windowsPowerShellModulePath, StringComparison.OrdinalIgnoreCase))
+                if (string.Equals(s_psModulePathEnvVar, currentModulePath, StringComparison.Ordinal) &&
+                    string.IsNullOrEmpty(InternalTestHooks.TestWindowsPowerShellPSHomeLocation) &&
+                    InternalTestHooks.TestPowerShellPSModulePaths?.Length == 0)
                 {
-                    newModulePathString += Path.PathSeparator + windowsPowerShellModulePath;
+                    // Another thread already processed %PSModulePath%.
+                    return s_modulePathCache;
                 }
+
+                // %PSModulePath% is changed so parse it in the thread and update module path cache.
+                s_modulePath = new List<string>();
```

Comment:
If this can be accessed by multiple threads, should it be a `ConcurrentCollection<string>`?

Reply:
It must be solid change otherwise we could get unpredictable results in module search.